lib/container_server.go

```diff
@@ -53,7 +54,41 @@ func (c *ContainerServer) Runtime() *oci.Runtime {
 
 // Hooks returns the oci hooks for the ContainerServer
 func (c *ContainerServer) Hooks() map[string]HookParams {
-	return c.hooks
+	hooks := map[string]HookParams{}
+	c.hooksLock.Lock()
+	for key, h := range c.hooks {
+		hooks[key] = h
+	}
+	c.hooksLock.Unlock()
+	return hooks
+}
+
+// RemoveHook removes an hook by name
+func (c *ContainerServer) RemoveHook(hook string) {
+	c.hooksLock.Lock()
+	defer c.hooksLock.Unlock()
+	_, ok := c.hooks[hook]
+	if ok {
+		delete(c.hooks, hook)
+	}
+}
+
+func (c *ContainerServer) AddHook(hookPath string) {
+	hook, err := readHook(hookPath)
+	if err != nil {
+		logrus.Debugf("error while reading hook %s", hookPath)
+		return
+	}
+	c.hooksLock.Lock()
+	defer c.hooksLock.Unlock()
```

Comment:
You did not use defer above and  you are using it here?   Is there a reason?  I thought there was some performance issues with defer?

Reply:
I guess it just style, I can remove the defer, performance isn't mesurable afaict

Reply:
I actually prefer the defer, but you did not use it above.  I prefer the defer since it protects you from some future person adding a return err into later code and we end up leaking.

Reply:
ok fixed